From 17259024b5abae2b49b607c5f0074e79f92e6aef Mon Sep 17 00:00:00 2001
From: wuzhouhui <wuzhouhui250@gmail.com>
Date: Fri, 16 Dec 2016 22:04:43 +0800
Subject: [PATCH 10/95] Paging output automatically

Previously, if we want svn's output to be paged, we have to use pipe, like
"svn log | less", now we have no longer to do this again, because svn can
paging output automatically, just like git (except color highlight).
---
 subversion/svn/svn.c | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/subversion/svn/svn.c b/subversion/svn/svn.c
index 952512d..67da467 100644
--- a/subversion/svn/svn.c
+++ b/subversion/svn/svn.c
@@ -29,6 +29,7 @@
 
 #include <string.h>
 #include <assert.h>
+#include <unistd.h>
 
 #include <apr_strings.h>
 #include <apr_tables.h>
@@ -3318,8 +3319,44 @@ main(int argc, const char *argv[])
 {
   apr_pool_t *pool;
   int exit_code = EXIT_SUCCESS;
+  pid_t pid;
+  int fd[2];
   svn_error_t *err;
 
+  if (pipe(fd) < 0) {
+    fprintf(stderr, "pipe failed\n");
+    exit(EXIT_FAILURE);
+  }
+  if ((pid = fork()) < 0) {
+    fprintf(stderr, "fork failed\n");
+    exit(EXIT_FAILURE);
+  } else if (pid == 0) { /* child */
+    close(fd[1]);
+    if (fd[0] != STDIN_FILENO) {
+      if (dup2(fd[0], STDIN_FILENO) != STDIN_FILENO) {
+        fprintf(stderr, "dup2 failed\n");
+        return EXIT_FAILURE;
+      }
+      close(fd[0]);
+    }
+
+    if (execlp("less", "less", "-F", "-X", (char *)0) < 0) {
+      fprintf(stderr, "exec failed\n");
+      exit(EXIT_FAILURE);
+    }
+    exit(EXIT_SUCCESS);
+  }
+
+  /* parent */
+  close(fd[0]);
+  if (fd[1] != STDOUT_FILENO) {
+    if (dup2(fd[1], STDOUT_FILENO) != STDOUT_FILENO) {
+      fprintf(stderr, "dup2 for stdout failed\n");
+      exit(EXIT_FAILURE);
+    }
+    close(fd[1]);
+  }
+
   /* Initialize the app. */
   if (svn_cmdline_init("svn", stderr) != EXIT_SUCCESS)
     return EXIT_FAILURE;
@@ -3341,6 +3378,8 @@ main(int argc, const char *argv[])
       svn_cmdline_handle_exit_error(err, NULL, "svn: ");
     }
 
+  close(STDOUT_FILENO);
+  waitpid(pid, NULL, 0);
   svn_pool_destroy(pool);
 
   svn_cmdline__cancellation_exit();
-- 
1.8.3.1

